internal/vcs/metadata.go

```diff
@@ -0,0 +1,244 @@
+		if len(pieces) != 2 {
+			return Metadata{}, fmt.Errorf("github basic auth credentials were malformed invalid auth components %+v", pieces)
+		}
+
+		headRef := gjson.GetBytes(event, "pull_request.head.ref").String()
+		clonePath := fmt.Sprintf("/tmp/%s", headRef)
+
+		// if the clone path already exists then let's just do a plain open. We might hit this
+		// if the user is running multiple Infracost commands on the head commit. We don't want
+		// to clone each time.
+		_, err = os.Stat(clonePath)
+		if err == nil {
+			r, err = git.PlainOpen(clonePath)
+			if err != nil {
+				return Metadata{}, fmt.Errorf("could open previously cloned path %w", err)
+			}
+		} else {
+			r, err = git.PlainClone(clonePath, false, &git.CloneOptions{
+				URL:           gjson.GetBytes(event, "repository.clone_url").String(),
+				ReferenceName: plumbing.NewBranchReferenceName(headRef),
+				Auth: &http.BasicAuth{
+					Username: pieces[0],
+					Password: pieces[1],
+				},
+				SingleBranch: true,
+				Depth:        1,
+			})
+			if err != nil {
+				return Metadata{}, fmt.Errorf("could not shallow clone github repo to fetchcommit information %w", err)
+			}
+		}
+
+		head, err := r.Head()
+		if err != nil {
+			return Metadata{}, fmt.Errorf("could not determine head from cloned github branch %w", err)
+		}
+
+		commit, err := r.CommitObject(head.Hash())
+		if err != nil {
+			return Metadata{}, fmt.Errorf("could not read head commit from cloned github repo %w", err)
+		}
+
+		m.Commit = commitToMetadata(commit)
+		m.Branch.Name = gjson.GetBytes(event, "pull_request.head.ref").String()
+	}
+
+	m.Pipeline = &Pipeline{ID: os.Getenv("GITHUB_RUN_ID")}
+	m.PullRequest = &PullRequest{
+		VCSProvider:  "github",
+		Title:        gjson.GetBytes(event, "pull_request.title").String(),
+		Author:       gjson.GetBytes(event, "pull_request.user.login").String(),
+		SourceBranch: gjson.GetBytes(event, "pull_request.head.ref").String(),
+		BaseBranch:   gjson.GetBytes(event, "pull_request.base.ref").String(),
+		URL:          gjson.GetBytes(event, "pull_request._links.html.href").String(),
+	}
+
+	return m, nil
+}
+
+func getLocalGitMetadata() (Metadata, error) {
+	r, err := git.PlainOpenWithOptions(".", &git.PlainOpenOptions{DetectDotGit: true})
+	if err != nil {
+		return Metadata{}, fmt.Errorf("could not open git directory to fetch metadata %w", err)
+	}
+
+	head, err := r.Head()
+	if err != nil {
+		return Metadata{}, fmt.Errorf("could not determine head from local git directory %w", err)
+	}
+
+	branch := head.Name().Short()
+	commit, err := r.CommitObject(head.Hash())
+	if err != nil {
+		return Metadata{}, fmt.Errorf("could not read head commit %w", err)
+	}
+
+	return Metadata{
+		Branch: Branch{Name: branch},
+		Commit: commitToMetadata(commit),
+	}, nil
+}
+
+func commitToMetadata(commit *object.Commit) Commit {
+	return Commit{
+		SHA:       commit.Hash.String(),
+		Author:    fmt.Sprintf("%s (%s)", commit.Author.Name, commit.Author.Email),
```

Comment:
Have you thought about splitting it into `AuthorName` and `AuthorEmail`? This might be handy if we want to display the author name nicely, support gravatar, etc.